[1] Shifu Chen, Tanxiao Huang, Yanqing Zhou, Yue Han, Mingyan Xu and Jia Gu. AfterQC: automatic filtering, trimming, error removing and quality control for fastq data. BMC Bioinformatics 2017 18(Suppl 3):80.
[2] Becker et al. FusorSV: an algorithm for optimally combining data from multiple structural variation detection methods. Genome Biology (2018) 19:38 doi.org/10.1186/s13059-018-1404-6
Blankenberg, D. and Coraor, N. and Von Kuster, G. and Taylor, J. and Nekrutenko, A. (2011). Integrating diverse databases into an unified analysis framework: a Galaxy approach. In Database, 2011 (0), pp. bar011–bar011. [doi:10.1093/database/bar011][Link]

Kent, W. J. and Sugnet, C. W. and Furey, T. S. and Roskin, K. M. and Pringle, T. H. and Zahler, A. M. and Haussler, a. D. (2002). The Human Genome Browser at UCSC. In Genome Research, 12 (6), pp. 996–1006. [doi:10.1101/gr.229102][Link]

Broad Institute. Picard. [Link]

Li, H. and Handsaker, B. and Wysoker, A. and Fennell, T. and Ruan, J. and Homer, N. and Marth, G. and Abecasis, G. and Durbin, R. (2009). The Sequence Alignment/Map format and SAMtools. In Bioinformatics, 25 (16), pp. 2078–2079. [doi:10.1093/bioinformatics/btp352][Link]

Quinlan, Aaron R. and Hall, Ira M. (2010). BEDTools: a flexible suite of utilities for comparing genomic features. In Bioinformatics, 26 (6), pp. 841–842. [doi:10.1093/bioinformatics/btq033][Link]

Li, H. and Durbin, R. (2009). Fast and accurate short read alignment with Burrows-Wheeler transform. In Bioinformatics, 25 (14), pp. 1754–1760. [doi:10.1093/bioinformatics/btp324][Link]

Li, Heng and Durbin, Richard (2010). Fast and accurate long-read alignment with Burrows–Wheeler transform. In Bioinformatics, 26 (5), pp. 589–595. [doi:10.1093/bioinformatics/btp698][Link]

Ramírez, Fidel and Ryan, Devon P and Grüning, Björn and Bhardwaj, Vivek and Kilpert, Fabian and Richter, Andreas S and Heyne, Steffen and Dündar, Friederike and Manke, Thomas (2016). deepTools2: a next generation web server for deep-sequencing data analysis. In Nucleic Acids Research, 44 (W1), pp. W160–W165. [doi:10.1093/nar/gkw257][Link]

Ewels, Philip and Magnusson, Måns and Lundin, Sverker and Käller, Max (2016). MultiQC: summarize analysis results for multiple tools and samples in a single report. In Bioinformatics, 32 (19), pp. 3047–3048. [doi:10.1093/bioinformatics/btw354][Link]

Heng Li (2013). Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. [Link]

Definition of SAM/BAM format. [Link]

Li, H. (2011). Improving SNP discovery by base alignment quality. In Bioinformatics, 27 (8), pp. 1157–1158. [doi:10.1093/bioinformatics/btr076][Link]

Li, H. (2011). A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data. In Bioinformatics, 27 (21), pp. 2987–2993. [doi:10.1093/bioinformatics/btr509][Link]

Danecek, P., Schiffels, S., Durbin, R.. Multiallelic calling model in bcftools (-m). [Link]

Durbin, R.. Segregation based metric for variant call QC. [Link]

Li, H.. Mathematical Notes on SAMtools Algorithms. [Link]

SAMTools GitHub page. [Link]

Erik Garrison (2012). Haplotype-based variant detection from short-read sequencing. [Link]

O. Tange (2011). GNU Parallel - The Command-Line Power Tool. In ;login: The USENIX Magazine, 36 (1), pp. 42-47. Frederiksberg, Denmark. [Link]
